EIPS/eip-4762.md

```diff
@@ -35,155 +35,266 @@ def get_storage_slot_tree_keys(storage_key: int) -> [int, int]:
 
+```python3
+def interpreter_loop(code: bytes, contract_address: bytes20):
+  pc = 0
+
+  # ...
+
+  while True:
+    if pc < len(code):
+      emit_code_access_events(pc, code, contract_address)
+      # ... rest of the interpreter loop
+
+# Executed when decoding an instruction when running the interpreter loop
+def emit_code_access_events(pc: int, code: bytes, contract_address: Address):
+  opcode = OP_STOP # pad chunk with 0s
+  # if the code walks beyond the code size, it is not considered to be accessed.
+  # XXX I don't think that this is true on mainnet, even before the atomic stuff!
+  if pc >= len(code):
+    return
+
+  opcode = code[pc]
+  chunk = pc // CHUNK_SIZE
+
+  add_read_access_event(contract_address, (chunk + CODE_OFFSET) // 256, (chunk + CODE_OFFSET) % 256)
+  charge_gas() # Note that the witness gas for instruction access is charged immediately,
+               # independently of access costs for its immediates.
+
+  # add access events for immediates, if the instruction has them
+  match opcode:
+    case OP_JUMP:
+      dest = stack.peek(0)
+      # The destination of a jump is not considered to be accessed if it is beyond the code
+      if dest < len(code):
+        # The destination of a `JUMP` is considered to be accessed, even if the destination is not a jumpdest or is inside pushdata
+        add_read_access_event(contract_address, (dest_chunk + CODE_OFFSET) // 256, (dest_chunk + CODE_OFFSET) % 256)
+
+    case OP_JUMPI:
+      dest = stack.peek(0)
+      cond = stack.peek(1)
+      # The destination of a jump is not considered to be accessed if it is beyond the code
+      if dest < len(code):
+        dest_chunk = dest // CHUNK_SIZE
+        # The destination of a positively evaluated `JUMPI` is considered to be accessed, even if the destination is not a jumpdest or is inside pushdata
+        # The destination of a `JUMPI` is not considered to be accessed if the jump conditional is `false`
+        if opcode == OP_JUMP or cond:
+          add_read_access_event(contract_address, (dest_chunk + CODE_OFFSET) // 256, (dest_chunk + CODE_OFFSET) % 256)
+
+    case OP_PUSH1 | OP_PUSH2 | ... | OP_PUSH32:
+      nbytes = opcode - OP_PUSH0 # determine the number of extra bytes to touch
+      for i in range(nbytes):
+        immediate_chunk = (pc+i) // CHUNK_SIZE
+        if immediate_chunk <= (len(code)+CHUNK_SIZE-1) // CHUNK_SIZE:
+          add_read_access_event(contract_address, (immediate_chunk + CODE_OFFSET) // 256, (immediate_chunk + CODE_OFFSET) % 256)
+
+    case OP_CODECOPY | OP_EXTCODECOPY:
+      addr = contract_address
+      size = 0
+      offset = 0
+      if opcode == OP_EXTCODECOPY:
+        addr = stack.peek(0)
+        offset = stack.peek(2)
+        size = stack.peek(3)
+      else:
+        offset = stack.peek(1)
+        size = stack.peek(2)
+
+      for i in range(size):
+        copied_byte_chunk = (offset+i) // CHUNK_SIZE 
+        if copied_byte_chunk <= (len(code)+CHUNK_SIZE-1) // CHUNK_SIZE:
```

Comment:
I think this is easier (removes the `- 1`)
```suggestion
        if copied_byte_chunk < (len(code)+CHUNK_SIZE) // CHUNK_SIZE:
```

`code` refers to locally loaded code which will not work for EXTCODECOPY on non-local addresses